### PR TITLE
Docs rewrite

### PR DESCRIPTION
**Pull request by daveads**
###  Roll out Comprehensive documentation for ZIO HTTP #2198 

Would like to keep experimenting and exploring zio-http to better understand what's going on and constantly update the doc...

*  Introduction ✅ 
*  Concepts 
     * Routing ✅ 
     * Request Handling ✅ 
     * Server ✅ 
     * Client ✅ 
     * Middleware ✅ 
     * Endpoint ✅ 
* Tutorials
     * Your First ZIO HTTP App ✅
     * Testing Your ZIO HTTP App
* How-to-guides
     * (This should be based off the list of topics at http4k, but should also include ZIO HTTP use cases around Endpoint, middleware, etc.) ✅ 
* Reference
     *  API Docs
     * Server backend
     * Websockets
     * JSON Handling 
     * Metrics
     * Request Logging
     * Example with explanations 
* Performance ✅ 
* FAQ ✅ 

/claim #2198

### CODE REVIEW

**Comment by daveads** [comment]
I need this drafted for now...

**Comment by daveads** [comment]
Made some changes to the **sidebar.js** so this run properly locally... after doing a read on the documentation 

For now my Major focus is on This... so should make a fast progress ASAP

File: README.md [hunk 1]
```diff
@@ -13,45 +13,170 @@ ZIO HTTP is a scala library for building http apps. It is powered by ZIO and [Ne
...
 
-**NOTES ON VERSIONING:**
+ZIO-HTTP provides a core library, `zhttp-http`, which includes the base HTTP implementation and server/client capabilities based on ZIO. Additional functionality like serverless support, templating, and websockets are available through separate add-on modules. ZIO-HTTP applications can be easily integrated into different deployment platforms, such as server-based, serverless, or compiled to native binaries.
+
+The principles of ZIO-HTTP are:
+
+- Application as a Function: HTTP services in ZIO-HTTP are composed of simple functions. The `HttpApp` type represents a function from an `HttpRequest` to a `ZIO` effect that produces an `HttpResponse`.
+- Immutability: Entities in ZIO-HTTP are immutable by default, promoting functional programming principles.
+- Symmetric: The same `HttpApp` interface is used for both defining HTTP services and making HTTP requests. This enables easy testing and integration of services without requiring an HTTP container.
+- Minimal Dependencies: The core `zhttp-http` module has minimal dependencies, and additional add-on modules only include dependencies required for specific functionality.
```
Comment (khajavi) [inline README.md]: `zhttp-http` -> `zio-http`

File: README.md [hunk 2]
```diff
@@ -13,45 +13,170 @@ ZIO HTTP is a scala library for building http apps. It is powered by ZIO and [Ne
...
+    Server.serve(app).provide(Server.default).exitCode
+}
 ```
 
-**NOTES ON VERSIONING:**
+ZIO-HTTP provides a core library, `zhttp-http`, which includes the base HTTP implementation and server/client capabilities based on ZIO. Additional functionality like serverless support, templating, and websockets are available through separate add-on modules. ZIO-HTTP applications can be easily integrated into different deployment platforms, such as server-based, serverless, or compiled to native binaries.
+
+The principles of ZIO-HTTP are:
+
+- Application as a Function: HTTP services in ZIO-HTTP are composed of simple functions. The `HttpApp` type represents a function from an `HttpRequest` to a `ZIO` effect that produces an `HttpResponse`.
```
Comment (khajavi) [inline README.md]: We do not have `HttpRequest` and `HttpResponse` data type anymore!

File: README.md [hunk 3]
```diff
@@ -27,31 +27,167 @@ A simple Http server can be built using a few lines of code.
...
+- Minimal Dependencies: The core `zhttp-http` module has minimal dependencies, and additional add-on modules only include dependencies required for specific functionality.
+- Testability: ZIO-HTTP supports easy in-memory and port-based testing of individual endpoints, applications, websockets/SSE, and complete suites of microservices.
+- Portability: ZIO-HTTP applications are portable across different deployment platforms, making them versatile and adaptable.
+
+By leveraging the power of ZIO and the simplicity of functional programming, ZIO-HTTP provides a robust and flexible toolkit for building scalable and composable HTTP services in Scala.
+
+## Quickstart
+
+Eager to start coding without delay? If you're in a hurry, you can follow the [quickstart](https://github.com/zio/zio-http/tree/main/zio-http-example) guide or explore the [examples repository](https://github.com/zio/zio-http/tree/main/zio-http-example), which demonstrates different use cases and features of ZIO-HTTP.
+
```
Comment (khajavi) [inline README.md]: Two links with the same content?

Comment (daveads) [reply README.md]: updated readme. it was outdated

File: README.md [hunk 4]
```diff
@@ -27,31 +27,167 @@ A simple Http server can be built using a few lines of code.
...
+## Module feature overview
```
Comment (khajavi) [inline README.md]: I am not a fan of creating content that closely mimics what is already available on http4k. Many of the contents under this section are not related to ZIO‌ HTTP.
I think it is better to focus on our strengths!

Comment (daveads) [reply README.md]: hmm most of the stuff i worked on is in the docs directory but i understand what you are saying focusing on zio-http strengths

File: docs/concepts/client.md [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+
+- **Response Handling**: Once a request is sent, the client receives a response from the server. ZIO HTTP allows us to handle the response using various combinators and methods. For example, we can extract the response body as a string, parse it as JSON, or process it as a stream of bytes.
+
+- **Client Configuration**: ZIO HTTP allows us to configure the behavior of the client through a `ZClient.Config` object. This configuration includes options such as request timeouts, follow redirects, SSL settings, and more.
+
+- **Dependency Injection**: ZIO's dependency injection capabilities enable us to provide the necessary dependencies to the client. This includes the client configuration, the client implementation, and other dependencies required for networking, DNS resolution, and other underlying functionality.
+
+- **Error Handling**: ZIO-HTTP provides comprehensive error handling mechanisms. Errors that may occur during the client interaction, such as network failures, timeouts, or invalid responses, are represented as typed errors in the ZIO effect. You can handle these errors using combinators like `catchAll`, `orElse`, or `fold`.
```
Comment (khajavi) [inline docs/concepts/client.md]: Use the consistent phrase "ZIO HTTP" instead of "ZIO-HTTP" across all the documents.

Comment (daveads) [reply docs/concepts/client.md]: okay

File: docs/concepts/endpoint.md [hunk 6]
```diff
@@ -0,0 +1,38 @@
+---
+id: endpoint
+title: Endpoint
+---
+
+Endpoints in ZIO HTTP represent individual API operations or routes that the server can handle. They encapsulate the structure and behavior of the API endpoints in a type-safe manner, making them fundamental building blocks for defining HTTP APIs.
```
Comment (khajavi) [inline docs/concepts/endpoint.md]: It is good to focus on the goal of `Endpoint` API‌ which tries to separate the definition of API from its implementation.

File: docs/concepts/endpoint.md [hunk 7]
```diff
@@ -0,0 +1,38 @@
...
+
+## Endpoint Definition
+
+Endpoints are defined using the `Endpoint` object's combinators, such as `get`, `post`, `path`, `query`, and more. These combinators allow us to specify various aspects of the endpoint, such as the HTTP method, URL path, query parameters, request/response bodies, and more.
+
+## Middleware
+
+Middleware can be applied to endpoints using the `@@` operator. Middleware allows us to add additional behavior or processing to the endpoint. For instance, we can handle authentication, validation, error handling, logging, or implement any custom logic needed for the endpoint using middleware.
```
Comment (khajavi) [inline docs/concepts/endpoint.md]: It is a good practice to enclose the data types inside backticks such as `Middleware`, `Response`, `Request`, `Route`, etc., wherever required throughout the entire documentation.

File: docs/concepts/endpoint.md [hunk 8]
```diff
@@ -0,0 +1,38 @@
...
+
+Inside the implementation function, we can use ZIO effects to perform computations, interact with dependencies, and produce the desired response. ZIO's functional approach makes it easy to handle errors, perform asynchronous operations, and compose complex behaviors within the endpoint implementation.
+
+## Endpoint Composition
+
+Endpoints can be composed together using operators like `++`, allowing us to build a collection of endpoints that make up our API. This composition enables us to structure the API by grouping related endpoints or creating reusable components.
+
+## Converting to App
+
+To serve the defined endpoints, they need to be converted to an HTTP application (`HttpApp`). This conversion is done using the `toApp` method, which prepares the endpoints to be served as an HTTP application.
```
Comment (khajavi) [inline docs/concepts/endpoint.md]: Please make sure this paragraph reflects the current implementation.

File: docs/concepts/endpoint.md [hunk 9]
```diff
@@ -0,0 +1,38 @@
+---
+id: endpoint
+title: Endpoint
```
Comment (khajavi) [inline docs/concepts/endpoint.md]: The content under this page needs a lot more attention. I think it does not reflect the exact point of the `Endpoint` concept.

Comment (daveads) [reply docs/concepts/endpoint.md]: okay

File: docs/concepts/endpoint.md [hunk 10]
```diff
@@ -0,0 +1,83 @@
...
+---
+
+---
+id: client
+title: Client
+---
```
Comment (khajavi) [inline docs/concepts/endpoint.md]: Why is there a client header on the endpoint.md file? Is it a mistake?

Comment (daveads) [reply docs/concepts/endpoint.md]: yea it is

File: docs/concepts/endpoint.md [hunk 11]
```diff
@@ -0,0 +1,78 @@
...
+id: endpoint
+title: Endpoint
+---
+
+Endpoints in ZIO HTTP are defined using the `Endpoint` object's combinators, which provide a type-safe way to specify various aspects of the endpoint. For instance, consider defining endpoints for retrieving user information and user posts:
+
+```scala
```
Comment (khajavi) [inline docs/concepts/endpoint.md]: Please make sure all of your code snippets are compile-safe [by adding `mdoc` modifiers](https://scalameta.org/mdoc/docs/modifiers.html).

Comment (daveads) [reply docs/concepts/endpoint.md]: noted

File: docs/concepts/endpoint.md (hunk 9 above)
Comment (khajavi) [reply docs/concepts/endpoint.md]: What is your plan for the concept section? Are you going to focus on the concepts of the endpoint article?

If so, why is the main focus not on its core idea? e.g.:
- What are its purposes and benefits?
- Why should I use the endpoint?
- Benefit of separating the definition of endpoint from its implementation?
- Doc and Client generation
- and some other important topics that can convey the importance of this data type

File: docs/concepts/endpoint.md [hunk 12]
```diff
@@ -0,0 +1,95 @@
+---
+id: endpoint
+title: Endpoint
+---
+
+Endpoints in ZIO HTTP are defined using the `Endpoint` object's combinators, which provide a type-safe way to specify various aspects of the endpoint. For instance, consider defining endpoints for retrieving user information and user posts:
+
+```scala mdoc
```
Comment (khajavi) [inline docs/concepts/endpoint.md]: Have you seen how this snippet will be rendered?

File: docs/concepts/endpoint.md [hunk 13]
```diff
@@ -0,0 +1,95 @@
...
+```
+
+In these examples, we use combinators like `Method.GET`, `int`, and `query` to define the HTTP method, URL path, path parameters, query parameters, and response types of the endpoints.
+
+### Middleware Application
+
+Middleware can be applied to endpoints using the `@@` operator to add additional behavior or processing. For example, we can apply authentication middleware to restrict access to certain endpoints:
+
+```scala mdoc:invisible
+
```
Comment (khajavi) [inline docs/concepts/endpoint.md:52]: Why invisible?

File: docs/concepts/endpoint.md [hunk 14]
```diff
@@ -0,0 +1,95 @@
...
+  }
+```
+
+In this example, the implementation function takes an `Int` representing the user ID and returns a ZIO effect that produces the same ID.
+
+### Endpoint Composition
+
+Endpoints can be composed together using operators like `++`, allowing us to build a collection of endpoints that make up our API:
+
+```scala mdoc:invisible
```
Comment (khajavi) [inline docs/concepts/endpoint.md]: also this?

Comment (daveads) [reply docs/concepts/endpoint.md]: sending an update

File: docs/concepts/endpoint.md (hunk 13 above)
Comment (daveads) [reply docs/concepts/endpoint.md:52]: removed

File: docs/concepts/endpoint.md (hunk 12 above)
Comment (daveads) [reply docs/concepts/endpoint.md]: yup

File: docs/concepts/endpoint.md (hunk 9 above)
Comment (daveads) [reply docs/concepts/endpoint.md]: This section outlines the ideas behind zio http.. 

using this as a guideline https://diataxis.fr/explanation/

Comment (daveads) [reply docs/concepts/endpoint.md]: This show more in depth to it.. https://github.com/zio/zio-http/blob/main/docs/dsl/endpoint.md

Comment (daveads) [reply docs/concepts/endpoint.md]: @khajavi

**Comment by daveads** [comment]
@jdegoes ready for review!

should push a better updates  Tutorial/ and more to How-to-guide/ before the end of the day

File: docs/concepts/endpoint.md [hunk 15]
```diff
@@ -0,0 +1,117 @@
...
+  }
+```
+
+Here, the `auth` middleware ensu authenticated users can access the `getUser` endpoint.
+
+### Endpoint Implementation
+
+Endpoints are implemented using the `implement` method, which takes a function specifying the logic to handle the request and generate the response. Inside the implementation function, ZIO effects can be used to perform computations and interact with dependencies:
+
+```scala
```
Comment (khajavi) [inline docs/concepts/endpoint.md]: Please make sure to include the `mdoc` modifier in all code snippets whenever possible.

Comment (daveads) [reply docs/concepts/endpoint.md]: sure

File: docs/concepts/endpoint.md [hunk 16]
```diff
@@ -0,0 +1,117 @@
...
+val getUser = Endpoint(Method.GET / "users" / int("userId")).out[Int]
+
+val getUserPosts = Endpoint(Method.GET / "users" / int("userId") / "posts" / int("postId"))
+  .query(query("name"))
+  .out[List[String]] @@ auth
+```
+
+In these examples, we use combinators like `Method.GET`, `int`, and `query` to define the HTTP method, URL path, path parameters, query parameters, and response types of the endpoints.
+
+### Middleware Application
```
Comment (khajavi) [inline docs/concepts/endpoint.md]: Why start with a header with level 3?
Its possible to start with level 2. something like this:

```md
## Level 2 (A Section)

## Level 2 (B Section)

### Level 3 (BA)

### Level 3 (BB)
```

File: docs/concepts/endpoint.md [hunk 17]
```diff
@@ -0,0 +1,117 @@
...
+
+The ZIO HTTP server requires an `HttpApp[R]` to run. The server can be started using the `Server.serve()` method, which takes the HTTP application as input and any necessary configurations:
+
+```scala 
+val run = Server.serve(app).provide(Server.default)
+```
+
+The server listens on the specified port, accepts incoming connections, and routes the incoming HTTP requests to the appropriate endpoints.
+
+## Purposes and Benefits of Endpoints in ZIO HTTP:
```
Comment (khajavi) [inline docs/concepts/endpoint.md]: One note I have here is that it is important to focus on the unique value of each concept before discussing common features.

When it comes to the `Endpoint` API, type safety should not be prioritized as it is not the primary objective of this data type. Instead, it aims to separate the API definition from its implementation. This leads us to have auto-generators:  tools that help developers automate the process of creating APIs, client-side code, and API documentation respectively.

Another note I have for you is to make this section concise and make it only in one section without the need for subsections like purposes, benefits, and why using endpoint.

File: docs/concepts/middleware.md [hunk 18]
```diff
@@ -0,0 +1,206 @@
+---
+id: middleware
+title: Middleware
+---
```
Comment (khajavi) [inline docs/concepts/middleware.md:4]: Can you please check if we have similar content on this page https://github.com/zio/zio-http/blob/main/docs/dsl/middleware.md
If so, how are you dealing with it?

File: docs/concepts/endpoint.md (hunk 17 above)
Comment (daveads) [reply docs/concepts/endpoint.md]: okay

File: docs/concepts/middleware.md [hunk 19]
```diff
@@ -0,0 +1,47 @@
...
+id: middleware
+title: Middleware in ZIO HTTP
+---
+
+In ZIO HTTP, middleware is a powerful concept that allows you to modify the behavior of HTTP requests and responses before and after they are processed by your application. Middleware functions are applied to the HTTP app using the `@@` operator, creating a pipeline of transformations that are executed in the order they are composed.
```
Comment (khajavi) [inline docs/concepts/middleware.md]: To ensure consistency in the documentation, please use the pronouns "us" and "ours".

```suggestion
In ZIO HTTP, middleware is a powerful concept that allows us to modify the behavior of HTTP requests and responses before and after they are processed by your application. Middleware functions are applied to the HTTP app using the `@@` operator, creating a pipeline of transformations that are executed in the order they are composed.
```

File: docs/concepts/middleware.md (hunk 18 above)
Comment (khajavi) [reply docs/concepts/middleware.md:4]: The newer version is better than the previous one. So if we focus on showing their core values, without taking into the details, it makes them better. It is good to mention about why they exist, what problems they solve, and talk a little about aspect-oriented programming and cross-cutting concerns.

Sometimes showing one/two (simple) examples (without going into details) is sufficient.

File: docs/concepts/middleware.md (hunk 19 above)
Comment (daveads) [reply docs/concepts/middleware.md]: okay

File: docs/concepts/middleware.md (hunk 18 above)
Comment (daveads) [reply docs/concepts/middleware.md:4]: okay

File: docs/concepts/endpoint.md [hunk 20]
```diff
@@ -1,117 +1,137 @@
...
+      HttpCodec.error[BookNotFound](Status.NotFound),
+      HttpCodec.error[AuthenticationError](Status.Unauthorized),
+    )
+```
 
-## Converting to App
+## OpenAPI Documentation
 
-To serve the defined endpoints, they need to be converted to an HTTP application (`HttpApp`). This conversion is done using the `toHttpApp` method:
+ZIO HTTP allows you to generate OpenAPI documentation from `Endpoint` definitions, which can be used to create Swagger UI routes.
```
Comment (khajavi) [inline docs/concepts/endpoint.md]:
```suggestion
ZIO HTTP allows us to generate OpenAPI documentation from `Endpoint` definitions, which can be used to create Swagger UI routes.
```

Comment (khajavi) [reply docs/concepts/endpoint.md]: Please do the same for the rest of the content.

File: docs/concepts/endpoint.md [hunk 21]
```diff
@@ -1,117 +1,137 @@
...
-```scala mdoc:silent
+ZIO HTTP supports handling multiple error types by using either `Either` or a common base class for all errors.
 
-val getUserPostsRoute =
-    getUserPosts.implement[Any] {
-      Handler.fromFunctionZIO[(Int, Int, String)] { case (id1: Int, id2: Int, query: String) =>
-        ZIO.succeed(List(s"API2 RESULT parsed: users/$id1/posts/$id2?name=$query"))
-      }
-    }
+```scala
```
Comment (khajavi) [inline docs/concepts/endpoint.md]: make sure the snippet is compile-safe using mdoc.

Comment (daveads) [reply docs/concepts/endpoint.md]: oh yea...

File: docs/concepts/endpoint.md [hunk 22]
```diff
@@ -1,117 +1,137 @@
 ---
 id: endpoint
-title: Endpoint
+title: Endpoints in ZIO HTTP
 ---
```
Comment (khajavi) [inline docs/concepts/endpoint.md:4]: render the docs and see the sidebar. It might not be what you expect!

File: docs/concepts/endpoint.md [hunk 23]
```diff
@@ -1,117 +1,137 @@
...
```
Comment (khajavi) [inline docs/concepts/endpoint.md:1]: I read your new effort, it became better than before.
One thing that I can note is that in concept pages we don't need detailed examples, we only want to show‌ (not practical detailed working) what that concept is and how that works, maybe with some simple examples. 
Please note that this comment doesn't mean you should change something, but I thought it might help you if you want to improve the quality of your writing.

Comment (daveads) [reply docs/concepts/endpoint.md:1]: noted

File: docs/concepts/endpoint.md (hunk 22 above)
Comment (daveads) [reply docs/concepts/endpoint.md:4]: i guess am leaving it with just `Endpoint`

**Comment by daveads** [comment]
@jdegoes Does having multiple PRs contradict this? /claim https://github.com/zio/zio-http/issues/2198
i would like to merge some part that are fully ready...

Managing, reviewing, updating with last changes, is kinda for reviewers and me.

**Comment by varshith257** [comment]
@khajavi Can I raise a PR? This issue lasted for a year and yet not been solved. I already started working on this and waiting for the response to raise a PR or not because of this raised PR. 

@daveads Can I know its progress and the reason for this long time? It would help me further

**Comment by khajavi** [comment]
> @khajavi Can I raise a PR? This issue lasted for a year and yet not been solved. I already started working on this and waiting for the response to raise a PR or not because of this raised PR.

@varshith257 Sure, please! It would also be helpful to coordinate with others to avoid doing duplicate work.

**Comment by varshith257** [comment]
Didn't get it. Let us wait for a response of @daveads or should I proceed futher

**Comment by khajavi** [comment]
@daveads 
As you have opted for smaller PRs, I will close this. However, if you wish to continue, feel free to reopen it.

**Comment by khajavi** [comment]
> Didn't get it. Let us wait for a response of @daveads or should I proceed futher

@varshith257 Do whatever you want, and yeah, please start! That was only a suggestion for not having duplication.

**Comment by varshith257** [comment]
> @varshith257 Do whatever you want, and yeah, please start! That was only a suggestion for not having duplication.

@khajavi Better looking for Colloboration to get it done intuitively and effectively which is a long awaited issue to be solved.
